Sources/Flatten.swift

```diff
@@ -341,25 +341,55 @@ extension SignalProtocol where Value: SignalProducerProtocol, Error == Value.Err
 			return disposable
 		}
 	}
-
+	
 	fileprivate func observeConcat(_ observer: Observer<Value.Value, Error>, _ disposable: CompositeDisposable? = nil) -> Disposable? {
-		let state = ConcatState(observer: observer, disposable: disposable)
+		let state = Atomic(ConcatState<Value.Value, Error>())
+		
+		func startNextIfNeeded() {
+			while let producer = state.modify({ $0.dequeue() }) {
+				producer.startWithSignal { signal, inner in
+					let handle = disposable?.add(inner) ?? nil
```

Comment:
If it's a nested optional I can see why it would be needed, but I'd solve that by declaring the type instead I think.